Add background network workload support

hildrum opened the pull request:
Simulate background network workload by sending traffic from a VM to itself via its external IP address.   (Part of supporting #436)

Sending from a VM to itself isn't exactly the same as sending from one VM to another, so I did some testing to determine if it looked close enough.

To determine whether the effect is likely to be similar in the self-to-self case and the self-to-other case, I compared the throughput of the iperf benchmark to a self-iperf benchmark.  If these two gave substantially different results, I'd conclude that modeling background network workload by sending from a VM to itself is a bad approach.  However, they look similar-enough to me.

| type | mean throughput | min | max | notes | 
|-------|---------|-------|-------| ---------------------|
| different | 1169 | 701 | 1346 | ten bi-directional runs, so 20 data points |
| self | 968 | 734 | 1216 |  ten runs, for a total of ten data points |

cmccoy commented on perfkitbenchmarker/vm_util.py:
```diff
@@ -77,6 +77,12 @@
+flags.DEFINE_integer('background_network_mbits', None,
+                     'Amount of background network traffic while '
+                     'running a benchmark')
+flags.DEFINE_integer('background_network_seconds', 600,
```
Can this be be the maximum acceptable value?

hildrum commented:
Removed the background_network_seconds option and replaced it with the largest 32 bit signed integer; added a background_network_iptype option which can either be `EXTERNAL` or `INTERNAL` to determine which ip address the background iperf sender will use.

skschneider commented on perfkitbenchmarker/configs/option_decoders.py:
```diff
@@ -83,6 +83,55 @@ def Decode(self, value, component_full_name, flag_values):
 
 
+class EnumDecoder(ConfigOptionDecoder):
+  """ Verifies that the config options value is on of the allowed set.
```
Spelling: "on" -> "one"

skschneider commented on perfkitbenchmarker/configs/option_decoders.py:
```diff
@@ -83,6 +83,55 @@ def Decode(self, value, component_full_name, flag_values):
+  Passes through the value unmodified
+  """
+
+  def __init__(self, option, none_ok=False, valid_values=[], **kwargs):
```
none_ok not required, since None can be put in valid_values to make it acceptable.

skschneider replied:
It's dangerous to use [] for the default value of a list. Python will create it once when the function is defined and re-use the same list instance for every call of the function when a value is not explicitly provided. In general, your intent is to create a new list each time the function is called, and the way to do that is by setting the default to None and initializing like so:
```
self.valid_values = valid_values or []
```
In this specific case, it may not make sense to have a default, considering that would mean the decoder doesn't accept any values.

skschneider commented on perfkitbenchmarker/configs/option_decoders.py:
```diff
@@ -83,6 +83,55 @@ def Decode(self, value, component_full_name, flag_values):
+    Args:
+      value: The value specified in the config.
+      component_full_name: string.  Fully qualified name of the
+          configurable compoennt containing the config option.
```
"compoennt" spelling

skschneider commented on perfkitbenchmarker/linux_virtual_machine.py:
```diff
@@ -551,11 +556,33 @@ def StartBackgroundWorkload(self):
       self.RemoteCommand(
           'nohup sysbench --num-threads=%s --test=cpu --cpu-max-prime=10000000 '
           'run 1> /dev/null 2> /dev/null &' % self.background_cpu_threads)
+    if self.background_network_mbits:
```
Here it's particularly noticeable that the functions are getting long, and for extensibility purposes we should consider what will happen if we add more types of background workloads. It would be ideal if we could move the code for each type of background task somewhere else instead of having a list of if-checks.

skschneider commented on perfkitbenchmarker/linux_virtual_machine.py:
```diff
@@ -551,11 +556,33 @@ def StartBackgroundWorkload(self):
+      self.AllowPort(BACKGROUND_IPERF_PORT)
+      self.RemoteCommand('nohup iperf --server --port %s &> /dev/null &' %
+                         BACKGROUND_IPERF_PORT)
+      (stdout, _) = self.RemoteCommand('pgrep iperf -n')
```
The first parentheses are unnecessary.

skschneider commented on perfkitbenchmarker/linux_virtual_machine.py:
```diff
@@ -551,11 +556,33 @@ def StartBackgroundWorkload(self):
+                         self.background_network_mbits))
+      iperf_cmd = 'nohup ' + base_iperf_cmd + ' &> /dev/null &'
+      self.RemoteCommand(iperf_cmd)
+      (stdout2, _) = self.RemoteCommand('pgrep iperf -n')
```
Same comment about parentheses. Personally, I find "stdout2" to be an awkward name. It looks like the first stdout from earlier does not need to be preserved beyond here, so why not reuse the variable?

skschneider commented on perfkitbenchmarker/vm_util.py:
```diff
@@ -77,6 +77,9 @@
                      'Number of threads of background cpu usage while '
                      'running a benchmark')
+flags.DEFINE_integer('background_network_mbits', None,
+                     'Amount of background network traffic while '
```
How about "Rate of background network traffic..."? It would be nice if the per-second component of the unit could also make its way into the flag name, but I can't think of an elegant way to do that since "mbps" could be megabytes per second.

hildrum replied:
Would prefer `background_mbits_per_sec` or `background_network_mbits_sec`?

skschneider replied:
I suppose it's more important to specify network. Beyond that, having the name read nicely is a plus. If ```background_network_mbits_per_sec``` works, that would be best, but if that's too long, ```background_network_mbits_sec``` would suffice for me.

skschneider commented on perfkitbenchmarker/vm_util.py:
```diff
@@ -95,6 +98,11 @@ class IpAddressSubset(object):
                   'the receiving VM is reachable by internal IP (REACHABLE), '
                   'external IP only (EXTERNAL) or internal IP only (INTERNAL)')
 
+flags.DEFINE_enum('background_network_iptype', IpAddressSubset.EXTERNAL,
```
I'd prefer "ip_type" over "iptype", but I'm unfamiliar if there's a standard.

cmccoy replied:
The equivalent for network benchmarks is `--ip_addresses`. Maybe `background_network_ip_address`?

hildrum replied:
I'm happy to change it to `background_network_ip_type`.  What's used for specifying ip address for benchmarks like ping and iperf is simply `ip_address`, but I didn't think `background_network_ip_address` would be as clear as `background_network_iptype`

skschneider replied:
```--ip_addresses``` strikes me as a confusing flag name, since it suggests the actual address values. I vote ```ip_type``` but don't feel strongly about this issue.

skschneider commented on perfkitbenchmarker/vm_util.py:
```diff
@@ -95,6 +98,11 @@ class IpAddressSubset(object):
+flags.DEFINE_enum('background_network_iptype', IpAddressSubset.EXTERNAL,
+                  (IpAddressSubset.INTERNAL, IpAddressSubset.EXTERNAL),
+                  'IP address type to use when generated background network '
+                  'workload')
```
"generating background network traffic"?

skschneider commented on tests/background_network_workload_test.py:
```diff
@@ -0,0 +1,245 @@
+# Copyright 2015 PerfKitBenchmarker Authors. All rights reserved.
```
2016

skschneider commented on tests/background_cpu_test.py:
```diff
@@ -12,7 +12,7 @@
 # limitations under the License.
 
-"""Tests for background workload"""
+"""Tests for background cpu workload & background workload framework"""
```
This is still just the cpu workload, right?

hildrum replied:
No, not exactly, the first couple of tests here are not CPU workload specific.  The `testBackgroundWorkloadSpec` doesn't test the background CPU specifically, just that the prepare/start/stop functions are called.

skschneider replied:
Ah, okay. Can those be separated? Like a different TestCase class or file?

skschneider commented on tests/background_network_workload_test.py:
```diff
@@ -0,0 +1,245 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Tests for background workload"""
```
background network workload

skschneider commented on tests/background_network_workload_test.py:
```diff
@@ -0,0 +1,245 @@
+  def testWindowsVMCausesError(self):
+    """ windows vm with background_network_mbits raises exception """
+    with mock_flags.PatchFlags() as mocked_flags:
+      self.setupCommonFlags(mocked_flags)
```
Since you're using the two lines above in every test function, consider using mock_flags.PatchTestCaseFlags in the class setUp function instead.

skschneider commented on tests/background_network_workload_test.py:
```diff
@@ -0,0 +1,245 @@
+      config = configs.LoadConfig(ping_benchmark.BENCHMARK_CONFIG, {}, NAME)
+      spec = benchmark_spec.BenchmarkSpec(config, NAME, UID)
+      spec.ConstructVirtualMachines()
+      with self.assertRaises(Exception):
```
Could you use a more specific exception type?

skschneider commented on tests/background_network_workload_test.py:
```diff
@@ -0,0 +1,245 @@
+      mocked_flags.os_type = benchmark_spec.WINDOWS
+      config = configs.LoadConfig(ping_benchmark.BENCHMARK_CONFIG, {}, NAME)
+      spec = benchmark_spec.BenchmarkSpec(config, NAME, UID)
+      spec.ConstructVirtualMachines()
```
These three lines above for creating and initializing a BenchmarkSpec also seem pretty common. Could they be in a helper function?

skschneider commented:
Took a pass.

hildrum commented:
PTAL.  I plan to add file background workload soon; in that pull request, I'll reorganize the too-long StartBackgroundWorkload, StopBackgroundWorkload, and PrepareBackgroundWorkload.

skschneider commented on perfkitbenchmarker/configs/option_decoders.py:
```diff
@@ -83,6 +83,50 @@ def Decode(self, value, component_full_name, flag_values):
+    self.valid_values = valid_values
+
+  def Decode(self, value, component_full_name, flag_values):
+    """ Verifies that the provided value is on the allowed list.
```
"on" -> "in"

skschneider commented on perfkitbenchmarker/linux_virtual_machine.py:
```diff
@@ -551,11 +556,33 @@ def StartBackgroundWorkload(self):
+                        (ip_address, BACKGROUND_IPERF_PORT,
+                         BACKGROUND_IPERF_SECONDS,
+                         self.background_network_mbits_per_sec))
+      iperf_cmd = 'nohup ' + base_iperf_cmd + ' &> /dev/null &'
```
Our style guide favors string formatting (```%``` or ```.format(...)```).

skschneider commented on tests/background_cpu_test.py:
```diff
@@ -49,11 +45,7 @@
-  def _CheckAndIncrement(self, throwaway=None, expected_last_call=None):
-    self.assertEqual(self.last_call, expected_last_call)
-    self.last_call += 1
+    super(TestBackgroundWorkload, self).setUp()
```
Since this function is only calling super, it can be removed.

skschneider commented on tests/background_network_workload_test.py:
```diff
@@ -0,0 +1,217 @@
+    self.mocked_flags.os_type = benchmark_spec.WINDOWS
+    config = configs.LoadConfig(ping_benchmark.BENCHMARK_CONFIG, {}, NAME)
+    spec = benchmark_spec.BenchmarkSpec(config, NAME, UID)
+    spec.ConstructVirtualMachines()
```
These lines above could be a call to makeSpec.

skschneider commented on tests/background_network_workload_test.py:
```diff
@@ -0,0 +1,217 @@
+
+  def testBackgroundWorkloadVanillaConfig(self):
+    """ Test that nothing happens with the vanilla config """
+    self.mocked_flags.background_network_mbits_per_sec = None
```
I think this line could be removed.

skschneider commented on tests/background_network_workload_test.py:
```diff
@@ -0,0 +1,217 @@
+  def testBackgroundWorkloadWindows(self):
+    """ Test that nothing happens with the vanilla config """
+    self.mocked_flags.os_type = benchmark_spec.WINDOWS
+    self.mocked_flags.background_network_mbits_per_sec = None
```
I think this line could also be removed.

skschneider commented on tests/background_network_workload_test.py:
```diff
@@ -0,0 +1,217 @@
+"""
+
+
+
```
Nit: Extra blank line. Usually have two.

skschneider commented on tests/background_network_workload_test.py:
```diff
@@ -0,0 +1,217 @@
+      self.assertEqual(vm.background_network_mbits_per_sec, 300)
+      self.assertEqual(vm.background_network_ip_type, 'INTERNAL')
+    self._CheckVMFromSpec(spec, 1)
+
```
Nit: Add another blank line here.

skschneider commented on tests/background_workload_framework_test.py:
```diff
@@ -0,0 +1,77 @@
+        self.assertEqual(vm.StartBackgroundWorkload.call_count, 1)
+        self.assertEqual(vm.StopBackgroundWorkload.call_count, 1)
+        self.assertEqual(vm.PrepareBackgroundWorkload.call_count, 1)
+
```
Nit: Add blank line.

skschneider commented:
Took another pass. Just a few style comments. Otherwise LGTM.
